AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/util/RetrofitExt.kt

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright 2021 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package dev.androidx.ci.util
+
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import okhttp3.Request
+import okio.Timeout
+import retrofit2.Call
+import retrofit2.CallAdapter
+import retrofit2.Callback
+import retrofit2.Response
+import retrofit2.Retrofit
+import java.lang.reflect.Type
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicBoolean
+
+@Target(AnnotationTarget.FUNCTION)
+annotation class Retry(val times: Int = 3)
+typealias RetryScheduler = (delay: Long, timeUnit: TimeUnit, block: suspend () -> Unit) -> Unit
+
+private class RetryCallAdapter(
+    val scheduler: RetryScheduler,
+    val delegate: CallAdapter<Any?, Any?>,
+    val retry: Int,
+) : CallAdapter<Any?, Any?> {
+    override fun responseType(): Type {
+        return delegate.responseType()
+    }
+
+    override fun adapt(call: Call<Any?>): Any {
+        return RetryCall(scheduler, delegate.adapt(call) as Call<Any?>, retry)
+    }
+}
+
+private class RetryCall(
+    private val scheduler: RetryScheduler,
+    private var delegate: Call<Any?>,
+    private val retryLimit: Int
+) : Call<Any?> {
+    private val cancelled = AtomicBoolean(false)
+    private var retryCount = 0
+    override fun clone(): Call<Any?> {
+        return RetryCall(scheduler, delegate.clone(), retryLimit)
+    }
+
+    override fun execute(): Response<Any?> {
+        error("sync ops are not supported yet")
+    }
+
+    override fun enqueue(callback: Callback<Any?>) {
+        if (isCanceled || isExecuted) {
+            return
+        }
+        delegate.enqueue(
+            RetryCallback(
+                delegate = callback,
+                shouldRetry = { response: Response<Any?>? ->
+                    retryCount < retryLimit && (
+                        response == null || response.code() >= 500
```

Comment:
Might want to consider making this more configurable (and HTTP 500 as default), so you can pass a lambda that can be evaluated if you need a retry. You might find it more useful especially with APIs where errors are not channeled using HTTP error codes.

Reply:
yea i thought about that but decided to keep it simple. Will do a followup if we see that happening.
Also, this still has not flaked, idk why :)
https://github.com/androidx/androidx/actions/workflows/integration_tests.yml
maybe because retrofit already retries connection failures and for the reset, backends have been stable enough.